// @roots/entrypoints-webpack-plugin/lib/html.emitter.js
import { __decorate } from "tslib";
import { bind } from 'helpful-decorators';
import Webpack from 'webpack';
/**
 * Emits inline html for each entrypoint
 *
 * @param compilation - Webpack compilation instance
 * @param publicPath - public path for entrypoints
 */
export class HtmlEmitter {
    compilation;
    assets;
    entrypoints;
    publicPath;
    /**
     * Class constructor
     *
     * @param compilation - webpack compilation
     * @param publicPath - asset publicPath
     */
    constructor(compilation, assets, entrypoints, publicPath) {
        this.compilation = compilation;
        this.assets = assets;
        this.entrypoints = entrypoints;
        this.publicPath = publicPath;
    }
    /**
     * Reduce entrypoint entrypoints to markup
     */
    emit() {
        ;
        [...this.entrypoints.entries()].map(([name, entrypoint]) => {
            this.compilation.emitAsset(`${name}.html`, new Webpack.sources.RawSource([...entrypoint.entries()].reduce(this.entrypointsReducer, ``)));
        });
    }
    /**
     * Reduce an entrypoint entry to markup
     */
    entrypointsReducer(acc, [type, files]) {
        if ([`js`, `mjs`].includes(type))
            return [...files].reduce(this.scriptReducer, acc);
        if (type === `css`)
            return [...files].reduce(this.styleReducer, acc);
        return acc;
    }
    /**
     * Get compiled asset file contents
     *
     * @param file - asset file
     * @returns - asset file contents
     */
    getCompiledAsset(file) {
        const raw = this.assets[file.replace(this.publicPath, ``)]?.source();
        return raw instanceof Buffer ? raw.toString() : raw;
    }
    /**
     * Reduce a js filename to markup
     */
    makeScript(attributes, inner = ``) {
        if (typeof inner !== `string`)
            return;
        inner = inner ? `\n\t${inner}\n` : ``;
        const stringyAttributes = attributes
            ? Object.entries(attributes)
                .filter(([, v]) => typeof v !== `undefined` && v !== false)
                .map(([k, v]) => (v === true ? k : `${k}=${v}`))
                .reduce((acc, v) => [...acc, v], [])
                .filter(Boolean)
                .join(` `)
            : ``;
        return `<script ${stringyAttributes}>${inner}</script>`;
    }
    /**
     * Reduce a script from entry item to markup
     */
    scriptReducer(acc, src) {
        const attributes = {
            async: true,
            defer: true,
            src: !src.includes(`runtime`) ? src : false,
            type: src.endsWith(`.mjs`) ? `module` : false,
        };
        return [
            acc,
            this.makeScript(attributes, src.includes(`runtime`) ? this.getCompiledAsset(src) : null),
        ].join(`\n`);
    }
    /**
     * Reduce a stylesheet from entry item to markup
     */
    styleReducer(acc, file) {
        return [acc, `<link rel=stylesheet href=${file} />`].join(`\n`);
    }
}
__decorate([
    bind
], HtmlEmitter.prototype, "emit", null);
__decorate([
    bind
], HtmlEmitter.prototype, "entrypointsReducer", null);
__decorate([
    bind
], HtmlEmitter.prototype, "getCompiledAsset", null);
__decorate([
    bind
], HtmlEmitter.prototype, "makeScript", null);
__decorate([
    bind
], HtmlEmitter.prototype, "scriptReducer", null);
__decorate([
    bind
], HtmlEmitter.prototype, "styleReducer", null);
