// @roots/bud-react/lib/extension/index.js
import { __decorate } from "tslib";
import { Extension } from '@roots/bud-framework/extension';
import { bind, dependsOn, expose, label, } from '@roots/bud-framework/extension/decorators';
import merge from '@roots/bud-support/lodash/merge';
/**
 * React configuration
 */
let BudReact = class BudReact extends Extension {
    /**
     * {@link Extension.configAfter}
     *
     */
    async boot(bud) {
        if (![this.useSWC, this.useTypeScript, this.useBabel].some(t => t === true)) {
            this.logger.warn(`No supported compiler found.`);
        }
        if (this.useSWC) {
            bud.swc.setJsc(merge(bud.swc.jsc, { transform: { react: { runtime: `automatic` } } }));
        }
        if (this.useBabel) {
            const babelPluginUrl = await this.resolve(`@babel/preset-react`, import.meta.url).catch(bud.catch);
            this.app.babel.setPreset(`@babel/preset-react`, babelPluginUrl);
        }
    }
    /**
     * Accessor for `@roots/bud-react/react-refresh`
     *
     * @readonly
     */
    get refresh() {
        return this.app.extensions.get(`@roots/bud-react/react-refresh`);
    }
    /**
     * Use babel
     *
     * @readonly
     */
    get useBabel() {
        if (this.useTypeScript)
            return false;
        if (this.useSWC)
            return false;
        return this.app.extensions.has(`@roots/bud-babel`);
    }
    /**
     * Use SWC
     *
     * @readonly
     */
    get useSWC() {
        return this.app.extensions.has(`@roots/bud-swc`);
    }
    /**
     * Use TypeScript
     *
     * @readonly
     */
    get useTypeScript() {
        if (this.useSWC)
            return false;
        if (this.app.extensions.has(`@roots/bud-typescript`)) {
            return !this.app.extensions.get(`@roots/bud-typescript`).get(`babel`);
        }
        return false;
    }
};
__decorate([
    bind
], BudReact.prototype, "boot", null);
BudReact = __decorate([
    label(`@roots/bud-react`),
    dependsOn([`@roots/bud-react/react-refresh`]),
    expose(`react`)
], BudReact);
export default BudReact;
