// @roots/bud-tailwindcss-theme-json/lib/extension.js
import { __decorate } from "tslib";
import { Extension } from '@roots/bud-framework/extension';
import { bind, label, options, } from '@roots/bud-framework/extension/decorators';
import WordPressJSONPlugin, {} from '@roots/wordpress-theme-json-webpack-plugin';
import * as tailwindAdapter from './tailwind/index.js';
/**
 * Support Tailwind values in {@link Bud.wpjson}
 */
let TailwindThemeJSON = class TailwindThemeJSON extends Extension {
    apply(compiler) {
        compiler.hooks.thisCompilation.tap(`@roots/bud-tailwindcss-theme-json`, compilation => {
            const hooks = WordPressJSONPlugin.getCompilationHooks(compilation);
            hooks.dependencies.tap(`@roots/bud-tailwindcss-theme-json`, dependencies => [...dependencies, this.app.tailwind.configPath]);
            hooks.options.tapPromise(`@roots/bud-tailwindcss-theme-json`, async (data) => {
                if (this.app.isDevelopment)
                    await this.app.tailwind.sourceConfig();
                if (this.options.colors) {
                    data = this.tapColorsManifestObject(data);
                }
                if (this.options.fontFamilies) {
                    data = this.tapFontFamiliesManifestObject(data);
                }
                if (this.options.fontSizes) {
                    data = this.tapFontSizesManifestObject(data);
                }
                return data;
            });
        });
    }
    /**
     * {@link Extension.register}
     */
    async register(bud) {
        if (!(`wpjson` in bud))
            return;
        bud.wpjson.useTailwindColors = this.useTailwindColors;
        bud.wpjson.useTailwindFontFamily = this.useTailwindFontFamily;
        bud.wpjson.useTailwindFontSize = this.useTailwindFontSize;
    }
    tapColorsManifestObject(options) {
        const palette = tailwindAdapter.palette.transform({
            ...this.app.tailwind.resolveThemeValue(`colors`, this.options.colorsExtendOnly),
        });
        return {
            ...(options ?? {}),
            settings: {
                ...(options?.settings ?? {}),
                color: {
                    ...(options?.settings?.color ?? {}),
                    palette,
                },
            },
            version: 2,
        };
    }
    tapFontFamiliesManifestObject(json) {
        const fontFamilies = tailwindAdapter.fontFamily.transform({
            ...this.app.tailwind.resolveThemeValue(`fontFamily`, this.options.fontFamiliesExtendOnly),
        });
        return {
            ...(json ?? {}),
            settings: {
                ...(json?.settings ?? {}),
                typography: {
                    ...(json?.settings?.typography ?? {}),
                    fontFamilies,
                },
            },
            version: 2,
        };
    }
    tapFontSizesManifestObject(options) {
        const fontSizes = tailwindAdapter.fontSize.transform({
            ...this.app.tailwind.resolveThemeValue(`fontSize`, this.options.fontSizeExtendOnly),
        });
        return {
            ...(options ?? {}),
            settings: {
                ...(options?.settings ?? {}),
                typography: {
                    ...(options?.settings?.typography ?? {}),
                    fontSizes,
                },
            },
            version: 2,
        };
    }
    /**
     * Use tailwind colors in theme.json
     */
    useTailwindColors(extendOnly = false) {
        this.app.wpjson.enable();
        this.set(`colors`, true);
        this.set(`colorsExtendOnly`, extendOnly);
        return this.app.wpjson;
    }
    /**
     * Use tailwind fontFamily in theme.json
     */
    useTailwindFontFamily(extendOnly = false) {
        this.app.wpjson.enable();
        this.set(`fontFamilies`, true);
        this.set(`fontFamiliesExtendOnly`, extendOnly);
        return this.app.wpjson;
    }
    /**
     * Use tailwind fontSize in theme.json
     */
    useTailwindFontSize(extendOnly = false) {
        this.app.wpjson.enable();
        this.set(`fontSizes`, true);
        this.set(`fontSizesExtendOnly`, extendOnly);
        return this.app.wpjson;
    }
};
__decorate([
    bind
], TailwindThemeJSON.prototype, "apply", null);
__decorate([
    bind
], TailwindThemeJSON.prototype, "register", null);
__decorate([
    bind
], TailwindThemeJSON.prototype, "tapColorsManifestObject", null);
__decorate([
    bind
], TailwindThemeJSON.prototype, "tapFontFamiliesManifestObject", null);
__decorate([
    bind
], TailwindThemeJSON.prototype, "tapFontSizesManifestObject", null);
__decorate([
    bind
], TailwindThemeJSON.prototype, "useTailwindColors", null);
__decorate([
    bind
], TailwindThemeJSON.prototype, "useTailwindFontFamily", null);
__decorate([
    bind
], TailwindThemeJSON.prototype, "useTailwindFontSize", null);
TailwindThemeJSON = __decorate([
    label(`@roots/bud-tailwindcss-theme-json`),
    options({
        colors: false,
        colorsExtendOnly: false,
        fontFamilies: false,
        fontFamiliesExtendOnly: false,
        fontSizes: false,
        fontSizesExtendOnly: false,
    })
], TailwindThemeJSON);
export { TailwindThemeJSON };
